# Materia: Aprendizaje Automático 1

# 1er Instancia Evaluativa  - Parte Práctica

# Consigna:  
    
Dado el DataSet winequality-white.csv (guardado en la carpeta **data**) ud deberá encontrar en MEJOR MODELO teniendo en cuenta que el tiempo de proceso y la capacidad de cómputo son de suma importancia. Para ello será necesario que encuente los 10 mejores resultados y que indique (a su entender) cual sería el modelo y los hiperparámetros mas óptimos segun la premisa y cual de esos 10 resultados sería el que preferiría utilizar para resolver el caso: 

## A tener en cuenta

- Analice los datos en profundidad, evalue si se trata de una problema de clasificaion o regresion y comente sus conclusiones. Toda descripcion debe estar demostrada con lineas de código que lo asistan en sus conclusiones. 

- Seleccionar el modelo que mejor **AC** obtenga sobre él. Aplique todos los modelos que conozca o crea conveniente utilizar.   

- Cuando sea necesario randomizar usar semilla = 123, para dividir dataset usar proporción 80/20. 

- Incluya un resumen con los hiperparámetros y resultados de evaluación del modelo elegido.

- No olvide que, dado que el resultado de su trabajo debe utilizarse en una aplicacion on line, el tiempo de procesamiento y al capacidad de cómputo son de suma importancia. Es por ello que se prefiere perder **precisión** en pos de **celeridad** en el resultado.

- Comente todo su código indicando lo que va haciendo y porque lo va haciendo. Recuerque que puede generar lineas de Markdown o utilizar el # para escribir sus comentarios en lineas de Código.

In [1]:
import numpy as np
import pandas as pd

## Carga de los datos:

Cargue los datos en un DataFrame, llámelo df:

In [2]:
df= pd.read_csv(r'C:\Users\Usuario\DOWNLOADS\winequality-white.csv', sep=';')
df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0              0.27         0.36            20.7      0.045   
1               6.3              0.30         0.34             1.6      0.049   
2               8.1              0.28         0.40             6.9      0.050   
3               7.2              0.23         0.32             8.5      0.058   
4               7.2              0.23         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  
0         8.8        6  
1         9.5        6  
2        10.1        6  
3         9.9        6  
4         9.9        6  
...       ...      ...  
4893     11.2        6  
4894      9.6        5  
4895      9.4        6  
4896     12.8        7  
4897     11.8        6  

[4898 rows x 12 columns]

In [3]:
df.quality.unique()

array([6, 5, 7, 8, 4, 3, 9], dtype=int64)

### De que tipo de problema se trata?

Podemos decir que es un problema de clasificacion ya que el dataframe al hyacer uso de la instruccion .unique() en la culumna quality solo puede dar 7 valores enteros, cuya relacion indica que son parte de una clasificacion.

In [4]:
# A continuacion separamos nuestro dataframe en x e y, siendo x las features e y el target, y luego dividiremos de nuevo a x e 
# y creando así los bloques train y test de cada uno.
from sklearn.model_selection import train_test_split
x = df.drop(['quality'], axis=1)
y = df.quality
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2, random_state=123)

In [5]:
# Con el x_train y el y_train, los vuelvo a dividir y genero los validations que seran 4: Val_x_train,Val_x_test,Val_y_train y 
# Val_y_test
Val_x_train,Val_x_test,Val_y_train,Val_y_test=train_test_split(x_train,y_train,test_size=0.2, random_state=123)

###  Creacion y entrenamiento de los modelos

Con los validations encontraré que modelo predice un mejor AC

#### Modelo de arbol de decision

In [174]:
# Creo el modelo de arbol de decision y lo entreno con los valores de val_x_train y val_y_train
from sklearn.tree import DecisionTreeClassifier
arbol=DecisionTreeClassifier(random_state=123, max_depth=100,max_features=0.5)
arbol.fit(Val_x_train,Val_y_train)

# Obtengo el Accuracy, para eso primero debo predecir valores de y solo con el val_x_test y luego a este resultado lo
# comparo con val_y_test con la instruccion accuracy_score
from sklearn.metrics import accuracy_score
y_pred= arbol.predict(Val_x_test)
acarbol=accuracy_score(Val_y_test,y_pred)
acarbol

0.5752551020408163

In [ ]:
arbolvaldf=pd.DataFrame()

In [ ]:
arbolvaldf

In [176]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
varCriterion= ['gini', 'entropy', 'log_loss']
varSplitter=['best', 'random']
varMax_depth = range(0,100)
VarMin_samples_split= range(2,10)
VarMin_samples_leaf= range(1,10)

VarMax_features=[range(0,11),'sqrt','log2']
VarMax_leaf_nodes=range(2,10)
VarMin_impurity_decrease= range(0,10)
acmax=0
lista=[]
for z in varCriterion:
    for y in varSplitter:
        for x in varMax_depth:
            for w in VarMin_samples_split:
                for v in VarMin_samples_leaf:
                    for t in VarMax_features:
                        for s in VarMax_leaf_nodes:
                            for r in VarMin_impurity_decrease:
                                # Creo el modelo de KNN y lo entreno con los valores de val_x_train y val_y_train
                                arbol=DecisionTreeClassifier(random_state=123, criterion=z,splitter=y,max_depth=x,min_samples_split=w,min_samples_leaf=v,min_weight_fraction_leaf=0.5,max_features=t,max_leaf_nodes=s,min_impurity_decrease=r)
                                arbol.fit(Val_x_train,Val_y_train)
                                # Obtengo el Accuracy, para eso primero debo predecir valores de y solo con el val_x_test y luego a 
                                #este resultado lo comparo con val_y_test con la instruccion accuracy_score
                                y_pred= arbol.predict(Val_x_test)
                                acarbol=accuracy_score(Val_y_test,y_pred)
                                acmax = ac
                                vecinomax=n
                                Weightsmax=j
                                Algorithmmax=x
                                Pmax=m
                                lista.append([acmax,vecinomax,Weightsmax,Algorithmmax,Pmax])
lista
knnvaldf=knnvaldf.append(lista, ignore_index=True)

TypeError: '>' not supported between instances of 'range' and 'float'

#### Modelo de KNN

Para encontrar los mejores 10 mejores hiperparametros a usar ys u respectivo accuracy, generaré un for anidado, con el que cada nivel de for contendrá una lista que contiene los posibles valores de los hiperparametros, para luego generar el modelo, probar cada uno de los hiperparametros.
Luego de crear el modelo con los hiperparametros, entrenará con los valores de val_x_train y val_y_train, luego predecirá los valores de val_x_test y comparará a estos valores obtenidos con los de val_y_test, generando así el accuracy del modelo usado. Todas las combinaciones posibles se guardarán en una lista.

Para encontrar los 10 mejores valores, generaré un DataFrame vacío y le cargaré las listas obtenidas, luego lo ordenaré por la columna de accuracy de orden descendente para así obtener los valores mas grandes.

In [91]:
knnvaldf=pd.DataFrame()

In [92]:
knnvaldf

Empty DataFrame
Columns: []
Index: []

In [93]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
n_range= range(5,100)
varWeights=["uniform", "distance"]
varAlgorithm = ["auto", "ball_tree", "kd_tree", "brute"]
VarP= [1,2]
acmax=0
lista=[]
for n in n_range:
    for j in varWeights:
        for x in varAlgorithm:
            for m in VarP:
                # Creo el modelo de KNN y lo entreno con los valores de val_x_train y val_y_train
                knn=KNeighborsClassifier(n_neighbors=n,weights=j, algorithm=x, p= m, leaf_size=30)
                knn.fit(Val_x_train,Val_y_train)
                # Obtengo el Accuracy, para eso primero debo predecir valores de y solo con el val_x_test y luego a 
                #este resultado lo comparo con val_y_test con la instruccion accuracy_score
                predknn_y=knn.predict(Val_x_test)
                ac= accuracy_score(Val_y_test,predknn_y)
                acmax = ac
                vecinomax=n
                Weightsmax=j
                Algorithmmax=x
                Pmax=m
                lista.append([acmax,vecinomax,Weightsmax,Algorithmmax,Pmax])
lista
knnvaldf=knnvaldf.append(lista, ignore_index=True)

In [94]:
n_range= range(100,200)
varWeights=["uniform", "distance"]
varAlgorithm = ["auto", "ball_tree", "kd_tree", "brute"]
VarP= [1,2]
acmax=0
lista=[]
for n in n_range:
    for j in varWeights:
        for x in varAlgorithm:
            for m in VarP:
                knn=KNeighborsClassifier(n_neighbors=n,weights=j, algorithm=x, p= m, leaf_size=30)
                knn.fit(Val_x_train,Val_y_train)
                predknn_y=knn.predict(Val_x_test)
                ac= accuracy_score(Val_y_test,predknn_y)
                acmax = ac
                vecinomax=n
                Weightsmax=j
                Algorithmmax=x
                Pmax=m
                lista.append([acmax,vecinomax,Weightsmax,Algorithmmax,Pmax])
lista
knnvaldf=knnvaldf.append(lista, ignore_index=True)

In [95]:
n_range= range(200,300)
varWeights=["uniform", "distance"]
varAlgorithm = ["auto", "ball_tree", "kd_tree", "brute"]
VarP= [1,2]
acmax=0
lista=[]
for n in n_range:
    for j in varWeights:
        for x in varAlgorithm:
            for m in VarP:
                knn=KNeighborsClassifier(n_neighbors=n,weights=j, algorithm=x, p= m, leaf_size=30)
                knn.fit(Val_x_train,Val_y_train)
                predknn_y=knn.predict(Val_x_test)
                ac= accuracy_score(Val_y_test,predknn_y)
                acmax = ac
                vecinomax=n
                Weightsmax=j
                Algorithmmax=x
                Pmax=m
                lista.append([acmax,vecinomax,Weightsmax,Algorithmmax,Pmax])
lista
knnvaldf=knnvaldf.append(lista, ignore_index=True)

In [96]:
n_range= range(300,400)
varWeights=["uniform", "distance"]
varAlgorithm = ["auto", "ball_tree", "kd_tree", "brute"]
VarP= [1,2]
acmax=0
lista=[]
for n in n_range:
    for j in varWeights:
        for x in varAlgorithm:
            for m in VarP:
                knn=KNeighborsClassifier(n_neighbors=n,weights=j, algorithm=x, p= m, leaf_size=30)
                knn.fit(Val_x_train,Val_y_train)
                predknn_y=knn.predict(Val_x_test)
                ac= accuracy_score(Val_y_test,predknn_y)
                acmax = ac
                vecinomax=n
                Weightsmax=j
                Algorithmmax=x
                Pmax=m
                lista.append([acmax,vecinomax,Weightsmax,Algorithmmax,Pmax])
lista
knnvaldf=knnvaldf.append(lista, ignore_index=True)

In [97]:
n_range= range(400,500)
varWeights=["uniform", "distance"]
varAlgorithm = ["auto", "ball_tree", "kd_tree", "brute"]
VarP= [1,2]
acmax=0
lista=[]
for n in n_range:
    for j in varWeights:
        for x in varAlgorithm:
            for m in VarP:
                knn=KNeighborsClassifier(n_neighbors=n,weights=j, algorithm=x, p= m, leaf_size=30)
                knn.fit(Val_x_train,Val_y_train)
                predknn_y=knn.predict(Val_x_test)
                ac= accuracy_score(Val_y_test,predknn_y)
                acmax = ac
                vecinomax=n
                Weightsmax=j
                Algorithmmax=x
                Pmax=m
                lista.append([acmax,vecinomax,Weightsmax,Algorithmmax,Pmax])
lista
knnvaldf=knnvaldf.append(lista, ignore_index=True)

In [100]:
knnvaldf.columns= ['acc','vecinomax','Weightsmax','Algorithmmax','Pmax']
knnvaldf.sort_values(by=['acc','vecinomax'], ascending=False).head(10)

acc  vecinomax Weightsmax Algorithmmax  Pmax
7912  0.576531        499   distance         auto     1
7914  0.576531        499   distance    ball_tree     1
7916  0.576531        499   distance      kd_tree     1
7918  0.576531        499   distance        brute     1
2888  0.576531        185   distance         auto     1
2890  0.576531        185   distance    ball_tree     1
2892  0.576531        185   distance      kd_tree     1
2894  0.576531        185   distance        brute     1
7896  0.575255        498   distance         auto     1
7898  0.575255        498   distance    ball_tree     1

Podemos apreciar que el mejor accuracy posible con el modelo de KNN es de 0.576531